src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,528 @@
+			}
+
+			/// <summary>
+			/// Unlock the context, causing <see cref="Wait"/> to cease blocking. Do nothing if the context
+			/// is not locked.
+			/// </summary>
+			protected virtual void Unlock()
+			{
+				// Do not set _locked back to false, we want to wait only once per transaction context.
+				// Volatile member, do not simplify by removing the local variable.
+				var semaphore = _semaphore;
+				_semaphore = null;
+				if (semaphore == null)
+					return;
+				// Not supposed to have more than one waiting thread, but in case the user made bad multi-threading,
+				// do not leave frozen threads.
+				semaphore.Release(100);
+				semaphore.Dispose();
+			}
+
+			#region IEnlistmentNotification Members
+
+			/// <summary>
+			/// Prepare the session for the transaction commit. Runs
+			/// <see cref="ISessionImplementor.BeforeTransactionCompletion(ITransaction)"/> for the session and for
+			/// <see cref="ConnectionManager.DependentSessions"/> if any. <see cref="Lock"/> the context
+			/// before signaling it is done, or before rollback in case of failure.
+			/// </summary>
+			/// <param name="preparingEnlistment">The object for notifying the prepare phase outcome.</param>
+			public virtual void Prepare(PreparingEnlistment preparingEnlistment)
+			{
+				using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+				{
+					try
+					{
+						using (_sessionImplementor.ConnectionManager.BeginsProcessingFromSystemTransaction(_useConnectionOnSystemTransactionEvents))
+						{
+							if (_useConnectionOnSystemTransactionEvents)
+							{
+								// Ensure any newly acquired connection gets enlisted in the transaction. When distributed,
+								// this code runs from another thread and we cannot rely on Transaction.Current.
+								using (var tx = new TransactionScope(AmbientTransaction))
+								{
+									_sessionImplementor.BeforeTransactionCompletion(null);
+									foreach (var dependentSession in _sessionImplementor.ConnectionManager.DependentSessions)
+										dependentSession.BeforeTransactionCompletion(null);
+
+									tx.Complete();
+								}
+							}
+							else
+							{
+								_sessionImplementor.BeforeTransactionCompletion(null);
+								foreach (var dependentSession in _sessionImplementor.ConnectionManager.DependentSessions)
+									dependentSession.BeforeTransactionCompletion(null);
+							}
+						}
+						// Lock the session to ensure second phase gets done before the session is used by code following
+						// the transaction scope disposal.
+						Lock();
+
+						_logger.Debug("Prepared and done for system transaction");
+						// We do not have anything more to do in second phases callbacks: the remaining work is handled in
+						// transaction completion event, which always executes, contrary to second phases callbacks which
+						// do not, depending on the rollback cases.
+						// This saves a thread when distributed.
+						preparingEnlistment.Done();
```

Comment:
> Can you show one of such cases?

The `ResourceManagerFixture.DistributedRollback(false, false, true)` demonstrates it. The volatile resource 3 has only its prepare called. Same for `NonDistributedRollback`.

`DistributedInDoubt` is another case, the durable resource 1 (support single phase but going 2PC in this case) has only its prepare phase called too.

> The code would be much simpler, because there will be no try-catch on getting wasSuccessful, there will be no unsubscription from the delegate, etc.

These are two small local points. What is etc? While having to cope with callbacks which are not always called depending on a number of conditions will surely be a bit less straightforward to handle than those two above points.

> They are. It's just a for loop.

Really? Check the results of the distributed tests of `ResourceManagerFixture`. Their logs are mixed. Sometimes they may look sequential, but just re-run them. Anyway, their thread numbers are not the same.

Like:

```
01:25:07,245 [21] - Durable resource 2: rollback phase, calling done
01:25:07,246 [23] - Single phase Durable resource 1: rollback phase end
01:25:07,248 [21] - Durable resource 2: rollback phase end
```